### PR TITLE
Deactivate Base64 Encoded Vso Commands AB#2008236

### PR DESCRIPTION
**Pull request by praval-microsoft**
**Issue:** Some pipeline variables may contain encoded vso commands which will be executed if it will be specified in script as output
We want to prevent this behavior by deactivating commands in these variables

**Description:** 
Added DeactivateBase64EncodedVsoCommands util which disables all vso commands in line

**Risk Assesment(Low/Medium/High)**: low

**Added unit tests:** (Y/N) Y

**Additional Tests Performed:** N

### CODE REVIEW

**Review by marekzmyslowski — DISMISSED** [review]

**Review by sanjuyadav24 — APPROVED** [review]

File: src/Agent.Sdk/Util/StringUtil.cs [hunk 1]
```diff
@@ -293,5 +293,28 @@ public static string DeactivateVsoCommands(string input)
...
+        public static string DeactivateBase64EncodedVsoCommands(string input)
+        {
+            if (string.IsNullOrEmpty(input))
+            {
+                return string.Empty;
+            }
+
+            try
+            {
+                byte[] decodedBytes = Convert.FromBase64String(input);
```
Comment (tarunramsinghani) [inline src/Agent.Sdk/Util/StringUtil.cs:306]: Add comments please. it is unclear what this is trying to achieve

File: src/Agent.Worker/WorkerUtilties.cs [hunk 2]
```diff
@@ -112,7 +112,7 @@ public static Pipelines.AgentJobRequestMessage DeactivateVsoCommandsFromJobMessa
                 if (deactivatedVariables.TryGetValue(variableName, out var variable))
                 {
                     var deactivatedVariable = variable ?? new VariableValue();
-
+                    deactivatedVariable.Value = StringUtil.DeactivateBase64EncodedVsoCommands(deactivatedVariable.Value);
```
Comment (tarunramsinghani) [inline src/Agent.Worker/WorkerUtilties.cs:115]: Better to put it inside the StringUtil.DeactivateVsoCommands to cover all the cases in case this func is used somewhere else as well.

File: src/Test/L0/Util/StringUtilL0.cs [hunk 3]
```diff
@@ -30,6 +32,49 @@ public void DeactivateVsoCommandsFromStringTest(string input, string expected)
             Assert.Equal(expected, result);
         }
 
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Common")]
+        public void DeactivateBase64EncodedVsoCommands_InputValidEncoded()
```
Comment (tarunramsinghani) [inline src/Test/L0/Util/StringUtilL0.cs:38]: Add comments on what this test cases and others are validating

File: src/Agent.Sdk/Util/StringUtil.cs [hunk 4]
```diff
@@ -293,5 +293,28 @@ public static string DeactivateVsoCommands(string input)
 
             return Regex.Replace(input, "##vso", "**vso", RegexOptions.IgnoreCase);
         }
+
+        public static string DeactivateBase64EncodedVsoCommands(string input)
+        {
+            if (string.IsNullOrEmpty(input))
+            {
+                return string.Empty;
```
Comment (AdityaMankal-MS) [inline src/Agent.Sdk/Util/StringUtil.cs:301]: Throw an exception if the input string is null or empty. See [InvalidArgumentException](https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.common.invalidargumentexception?view=sql-smo-160)

File: src/Agent.Sdk/Util/StringUtil.cs [hunk 5]
```diff
@@ -293,5 +293,28 @@ public static string DeactivateVsoCommands(string input)
...
+            {
+                byte[] decodedBytes = Convert.FromBase64String(input);
+                string decodedString = Encoding.UTF8.GetString(decodedBytes);
+
+                decodedString = DeactivateVsoCommands(decodedString);
+
+                byte[] bytes = Encoding.UTF8.GetBytes(decodedString);
+                return Convert.ToBase64String(bytes);
+            }
+            catch
```
Comment (AdityaMankal-MS) [inline src/Agent.Sdk/Util/StringUtil.cs:314]: Log the exception and throw it. Do not eat the exception without leaving any trace of it. A util method should leave it up to the caller to decide what to do with the exception.

This implementation leaves no trace of the exception and makes debugging (or even figuring out that there is an exception) very challenging

File: src/Test/L0/Util/StringUtilL0.cs [hunk 6]
```diff
@@ -30,6 +32,49 @@ public void DeactivateVsoCommandsFromStringTest(string input, string expected)
...
+        {
+            String vsoCommand = "##vso[task.setvariable variable=downloadUrl]https://www.evil.com";
```
Comment (AdityaMankal-MS) [inline src/Test/L0/Util/StringUtilL0.cs:40]: Use the keyword `string`. A reference to class `System.String` is not needed. Please do the same correction for all such declarations.

File: src/Test/L0/Util/StringUtilL0.cs [hunk 7]
```diff
@@ -30,6 +32,49 @@ public void DeactivateVsoCommandsFromStringTest(string input, string expected)
...
+            String encodedVsoCommand = Convert.ToBase64String(Encoding.UTF8.GetBytes(vsoCommand));
+
+            var result = StringUtil.DeactivateBase64EncodedVsoCommands(encodedVsoCommand);
```
Comment (AdityaMankal-MS) [inline src/Test/L0/Util/StringUtilL0.cs:43]: Do not use `var` when the right-hand side of an assignment is a function call. It reduces the readability.

Tough to understand what the return type is.
var someVariable = someFunctionCall()

Easy to understand what the return type is.
string someVariable = someFunctionCall()

Please do the same correction for all such cases.

File: src/Test/L0/Util/StringUtilL0.cs (hunk 3 above)
Comment (AdityaMankal-MS) [inline src/Test/L0/Util/StringUtilL0.cs:38]: Test names should follow this pattern 'MethodName_{Scenario}_{Expectation}`

Example: `DeactivateBase64EncodedVsoCommands_EncodedVsoCommands_Returns_DeactivatedVsoCommands`

Please do the same correction for all tests

File: src/Test/L0/Util/StringUtilL0.cs [hunk 8]
```diff
@@ -30,6 +32,49 @@ public void DeactivateVsoCommandsFromStringTest(string input, string expected)
...
+        [Trait("Category", "Common")]
+        public void DeactivateBase64EncodedVsoCommands_InputEmpty()
+        {
+            String vsoCommand = "";
+
+            var result = StringUtil.DeactivateBase64EncodedVsoCommands(vsoCommand);
+
+            Assert.Equal(vsoCommand, result);
+        }
+
```
Comment (AdityaMankal-MS) [inline src/Test/L0/Util/StringUtilL0.cs:77]: Add an additional test to cover null argument being passed to the method.

File: src/Test/L0/Worker/WorkerL0.cs [hunk 9]
```diff
@@ -293,13 +294,33 @@ public void VerifyJobRequestMessageVsoCommandsDeactivatedIfVariableCasesHandlesN
...
             Assert.Equal("", scrubbedMessage.Variables[Constants.Variables.Build.SourceVersionMessage]);
             Assert.Equal("", scrubbedMessage.Variables[Constants.Variables.System.SourceVersionMessage]);
-            Assert.Equal(" ", scrubbedMessage.Variables[Constants.Variables.Build.DefinitionName]);
+            Assert.Equal("", scrubbedMessage.Variables[Constants.Variables.Build.DefinitionName]);
+        }
+
+        [Fact]
+        [Trait("Level", "L0")]
+        [Trait("Category", "Worker")]
+        public void VerifyJobRequestMessageVsoBase64EncodedCommandsDeactivated()
```
Comment (AdityaMankal-MS) [inline src/Test/L0/Worker/WorkerL0.cs:309]: Please see the comments added to `StringUtilL0.cs`. Same corrections apply to this PR as well.

File: src/Agent.Sdk/Util/StringUtil.cs [hunk 10]
```diff
@@ -293,5 +293,28 @@ public static string DeactivateVsoCommands(string input)
...
         }
+
+        public static string DeactivateBase64EncodedVsoCommands(string input)
+        {
+            if (string.IsNullOrEmpty(input))
+            {
+                return string.Empty;
+            }
+
+            try
```
Comment (AdityaMankal-MS) [inline src/Agent.Sdk/Util/StringUtil.cs:304]: High number of variable assignments. Consider reducing them to a degree which doesn't drastically affect code readability

**Review by AdityaMankal-MS — CHANGES_REQUESTED** [review]

File: src/Agent.Sdk/Util/StringUtil.cs (hunk 4 above)
Comment (praval-microsoft) [reply src/Agent.Sdk/Util/StringUtil.cs:310]: Input can be null or empty here.

File: src/Agent.Sdk/Util/StringUtil.cs [hunk 11]
```diff
@@ -291,6 +292,39 @@ public static string DeactivateVsoCommands(string input)
...
+        {
+
+            if (string.IsNullOrEmpty(input))
+            {
+                return string.Empty;
+            }
+
+            try
+            {
+                string decodedString = Encoding.UTF8.GetString(Convert.FromBase64String(input));
```
Comment (tarunramsinghani) [inline src/Agent.Sdk/Util/StringUtil.cs:315]: Better to use [Convert.TryFromBase64String](https://learn.microsoft.com/en-us/dotnet/api/system.convert.tryfrombase64string?view=net-8.0) and use retrun value to decide instead of depending on exception. Catch(exception) is very generic and may cause unneccesary issues.

File: src/Agent.Sdk/Util/StringUtil.cs (hunk 4 above)
Comment (AdityaMankal-MS) [reply src/Agent.Sdk/Util/StringUtil.cs:310]: Yes. So when it is null or empty, throw an [InvalidArgumentException](https://learn.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.common.invalidargumentexception?view=sql-smo-160)

File: src/Agent.Sdk/Util/StringUtil.cs [hunk 12]
```diff
@@ -291,6 +292,43 @@ public static string DeactivateVsoCommands(string input)
...
+                throw new ArgumentNullException(nameof(input), "Input string cannot be null.");
+            }
+
+            if (input.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            byte[] buffer = new byte[input.Length];
+            if (Convert.TryFromBase64String(input, buffer, out int bytesWritten))
```
Comment (AdityaMankal-MS) [inline src/Agent.Sdk/Util/StringUtil.cs:318]: TryFromBase64String should be used when we are not always certain if the conversion would work. However, you are authoring a utility method called `DeactivateBase64EncodedVsoCommands`. The caller would have no clue if the conversion failed as you are returning the input as is when the conversion fails. Furthermore, failure of this conversion is not even logged making debugging even more difficult.

It is suggested to attempt conversion normally and throw the exception from this method. Let the caller decide what they want to do with it. Either they handle it or terminate the flow depending on their scenario.

**Review by AdityaMankal-MS — CHANGES_REQUESTED** [review]

File: src/Agent.Sdk/Util/StringUtil.cs (hunk 12 above)
Comment (praval-microsoft) [reply src/Agent.Sdk/Util/StringUtil.cs:318]: Updated the function name

**Review by tarunramsinghani — APPROVED** [review]

File: src/Agent.Sdk/Util/StringUtil.cs (hunk 12 above)
Comment (AdityaMankal-MS) [reply src/Agent.Sdk/Util/StringUtil.cs:325]: Glad that the exception is being thrown instead of returning the input as is when the conversion fails. And here is a simplified version of the function. 

```
  public static string DeactivateVsoCommandsIfBase64Encoded(string input)
      {
          if (input == null)
          {
              throw new ArgumentNullException(nameof(input), "Input string cannot be null.");
          }
          if (input.Length == 0)
          {
              return string.Empty;
          }
          byte[] decodedBytes = Convert.FromBase64String(input);
          string decodedString = Encoding.UTF8.GetString(decodedBytes);
          decodedString = ScrapVsoCommands(decodedString);
          return Convert.ToBase64String(Encoding.UTF8.GetBytes(decodedString));
      }
```

**Review by AdityaMankal-MS — APPROVED** [review]
Approved with suggestions

**Review by tarunramsinghani — APPROVED** [review]
